test/zipkin/util.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+//zipkin_util has constants and helper methods useful for zipkin tracing support.
+
+package zipkin
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"strconv"
+
+	"github.com/knative/pkg/test/logging"
+	"go.opencensus.io/trace"
+	"k8s.io/client-go/kubernetes"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+
+	// ZipkinPortForwardPid environment variable name to store Process ID of the process started for zipkin port-foward.
+	ZipkinPortForwardPid = "ZIPKIN_PORTFWD_PID"
+
+	//ZipkinTraceIDHeader HTTP response header key to be used to store Zipkin Trace ID.
+	ZipkinTraceIDHeader = "ZIPKIN_TRACE_ID"
+
+	// ZipkinPort port exposed by the Zipkin Pod
+	// https://github.com/knative/serving/blob/master/config/monitoring/200-common/100-zipkin.yaml#L25 configures the Zipkin Port on the cluster.
+	ZipkinPort = 9411
+
+	// ZipkinTraceEndpoint port-forwarded zipkin endpoint
+	ZipkinTraceEndpoint = "http://localhost:9411/api/v2/trace/"
+
+	// ZipkinNamespace namespace where zipkin pod runs.
+	// https://github.com/knative/serving/blob/master/config/monitoring/200-common/100-zipkin.yaml#L19 configures the namespace for zipkin.
+	ZipkinNamespace = "istio-system"
+)
+
+// SetupZipkinTracing sets up zipkin tracing which involves a) Setting up port-forwarding from localhost to zipkin pod on the cluster (pid of the process doing Port-Forward is stored in an env variable).
+// b) enable AlwaysSample config for tracing.
+func SetupZipkinTracing(kubeClientset *kubernetes.Clientset) error {
+
+	logger := logging.GetContextLogger("SpoofUtil")
+
+	zipkinPortAvailable, err := CheckZipkinPortAvailability()
+	if !zipkinPortAvailable {
+		return fmt.Errorf("Zipkin port not available on the machine: %v. Zipkin Setup failed", err)
+	}
+
+	zipkinPods, err := kubeClientset.CoreV1().Pods(ZipkinNamespace).List(metav1.ListOptions{LabelSelector: "app=zipkin"})
+
+	if err != nil {
+		return fmt.Errorf("Error retrieving Zipkin pod details : %v. Zipkin Setup failed", err)
+	}
+
+	if len(zipkinPods.Items) == 0 {
+		return errors.New("No Zipkin Pod found on the cluster. Ensure monitoring is switched on for your Knative Setup. Zipkin Setup failed")
```

Comment:
I dont think we need to add Zipkin setup failed to the error messages. The trace will metion the function name SetupZipkin

Reply:
Fixed.